test/integration/targets/package_facts/tasks/main.yml

```
@@ -18,6 +18,34 @@
     - name: check for ansible_facts.packages exists
       assert:
         that: ansible_facts.packages is defined
+
+    - name: Now try again but installing misleading rpm
+      block:
+        - name: install misleading rpm api
+          package: name="python3-rpm" state=present
+
+        - name: prep outputdir
+          tempfile: path=~ state=directory
+          register: tempdir
+
+        - name: install misleading rpm 'binary' file
+          file: dest="{{tempdir['path']}}/rpm" state=touch mode='0700'
+
+        - name: Gather package facts, finding 'rpm' on debian fam (needed for latest version)
+          package_facts:
```

The order from `auto` doesn't seem to be constant, so you could list the managers to make sure rpm is first.

```suggestion
          package_facts:
            managers:
              - rpm
              - apt
```

---

i specifically want to test auto, will check why order is not kept, weird since it is a list

---

I backed out the fix to run the tests, and at least the several times I ran them, couldn't trigger the bug (but could with the hardcoded managers list). It looked like the order was determined by https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/facts/packages.py#L17-L19.

---

there is a 'set' in get_all_subclasses that breaks predictability

---

The code change still lgtm. I think the test needs to exercise a broken package manager alphabetically preceding the working one. As written, the three setup tasks for this test are inconsequential since package_facts stops after apt succeeds for `strategy: first` (and it doesn't look like `strategy: all` could reproduce the bug). `manager: ['rpm', 'apt']` doesn't exercise `auto`, but it would force the bug being fixed.

---

added